scipy/sparse/construct.py

```diff
@@ -785,7 +793,27 @@ def data_rvs(n):
         else:
             data_rvs = random_state.rand
 
-    ind = random_state.choice(mn, size=k, replace=False)
+    # Number of non zero values
+    k = int(density * m * n)
+    if density_estimation == "exact":
+        ind = random_state.choice(mn, size=k, replace=False)
+    elif density_estimation == "probabilistic":
+        if k >= n*m:
+            raise ValueError("expected density < 1 with "
+                             "density_estimation='probabilistic'!")
+
+        # When sampling (1..N) with replacement p times we get
+        # on average k unique values,
+        #     k = N*(1 - ((N - 1)/N)**p)
```

Comment:
Formula taken from https://math.stackexchange.com/a/72351/371263. I have not double checked the math, but on the few examples, I tried it worked reasonably well.